### PR TITLE
Project/#1 create index

### PR DESCRIPTION
**Pull request by simonstadlinger**
@mweisgut This is the new pull request we are working on for now

### CODE REVIEW

**Comment by mweisgut** [comment]
```
git submodule update --init --recursive
```
results in
```
fatal: remote error: upload-pack: not our ref 4e37b385e7220a8cf934f39efb4a98cc765ce153
Fetched in submodule path 'third_party/sql-parser', but it did not contain 4e37b385e7220a8cf934f39efb4a98cc765ce153. Direct fetching of that commit failed.
```
 
Do you perhaps use a commit hash that has not been pushed to the remote repository?

**Comment by mweisgut** [comment]
In addition, the branch is 12 commits behind hyrise:master. Could you please sync it with hyrise:master, i.e., merge hyrise:master into this branch?

File: src/test/lib/ddl/ddl_statement_test.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+class DDLStatementTest : public BaseTest {
+ protected:
+  void SetUp() override {
+    Hyrise::reset();
+
+    // We reload table_a every time since it is modified during the test case.
+    _table_a = load_table("resources/test_data/tbl/int_float_create_index_test.tbl", 2);
+    Hyrise::get().storage_manager.add_table("table_a", _table_a);
+    auto compression_task_0 = std::make_shared<ChunkCompressionTask>("table_a", ChunkID{0});
+    auto compression_task_1 = std::make_shared<ChunkCompressionTask>("table_a", ChunkID{1});
```
Comment (mweisgut) [inline src/test/lib/ddl/ddl_statement_test.cpp]: You could encode the table first using `ChunkEncoder::encode_all_chunks(_table_a)` and add the encoded table to the storage manager afterward. Is there a certain reason why you want to use `ChunkCompressionTask`s after adding the table to the storage manager?

File: src/test/lib/ddl/ddl_statement_test.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  auto sql_pipeline = SQLPipelineBuilder{_create_index}.create_pipeline();
+
+  const auto& [pipeline_status, table] = sql_pipeline.get_result_table();
+  EXPECT_EQ(pipeline_status, SQLPipelineStatus::Success);
+
+  auto targeted_table = Hyrise::get().storage_manager.get_table("table_a");
+  auto column_ids = std::make_shared<std::vector<ColumnID>>();
+  column_ids->emplace_back(ColumnID{0});
+
+  auto actual_index = targeted_table->indexes_statistics().at(0);
```
Comment (mweisgut) [inline src/test/lib/ddl/ddl_statement_test.cpp]: Besides checking the index statistics, please also check the existence of the created index structures. Indexes are stored in a vector of `AbstractIndex` shared pointers in chunk objects (see the member variable `Chunk::_indexes`). `AbstractIndex::is_index_for` might be helpful to check if existing indexes are created for the correct segment(s).

File: src/lib/operators/maintenance/create_index.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  }
+  stream << ")";
+
+  return stream.str();
+   */
+  return std::string("create index description");
+}
+
+std::shared_ptr<const Table> CreateIndex::_on_execute(std::shared_ptr<TransactionContext> context) {
+  auto table_to_be_indexed = std::const_pointer_cast<Table>(left_input()->get_output());
```
Comment (mweisgut) [inline src/lib/operators/maintenance/create_index.cpp]: If you check the description of the `GetTable` operator, you will find the following information:
>The returned table is NOT
// the same table as stored in the StorageManager. If that stored table is changed (most importantly: if a chunk is
// added), this is not reflected in GetTable's result.

In addition, changes that you apply on the output table of the `GetTable` operator are not applied on the base table that is stored in the `StorageManager`. In other words: we need to get the table from the storage manager during the execution of the `CreateIndex` operator. Therefore, we actually don't need the output of the `GetTable` operator. We thus can omit the `GetTable` operator and pass the required information, i.e., the table name, directly to the `CreateIndex` operator.

File: src/test/lib/ddl/ddl_statement_test.cpp (hunk 1 above)
Comment (simonstadlinger) [reply src/test/lib/ddl/ddl_statement_test.cpp]: We incorporated that now! We also updated create index, create index node, drop index, and drop index node tests accordingly

File: src/test/lib/ddl/ddl_statement_test.cpp (hunk 2 above)
Comment (simonstadlinger) [reply src/test/lib/ddl/ddl_statement_test.cpp]: Thank you for the hint. `_indexes` is a private variable, so we built our check the other way round, which should be also sufficient.

File: src/lib/operators/maintenance/create_index.cpp (hunk 3 above)
Comment (simonstadlinger) [reply src/lib/operators/maintenance/create_index.cpp]: Thank you for the hint, we refactored it, and it works now! 🔥

**Comment by simonstadlinger** [comment]
@mweisgut Could you please have a look at the [DDLStatementTest-CreateIndexIfNotExistsAbort](https://github.com/simonstadlinger/hyrise/blob/project/%231-create-index/src/test/lib/ddl/ddl_statement_test.cpp)? The operator is supposed to fail because the index with the name already exists, but the logic error can not be caught for some reason. In the `create_index_test.cpp`, we have access to the context and can roll it back, which is okay at that level, but this should happen automatically when working with the high-level SQL pipeline, shouldn't it? Regardless of that, we don't know how to test that properly

File: src/test/lib/ddl/ddl_statement_test.cpp (hunk 2 above)
Comment (mweisgut) [reply src/test/lib/ddl/ddl_statement_test.cpp]: Verifying that the index statistics were created does not verify the successful creation of the index structures.

If private or protected members are to be tested, the test class can be declared as `friend class` (see https://en.cppreference.com/w/cpp/language/friend). An example can be seen in the `SettingsManager` where `SettingsManagerTest` is the friend class. Therefore, the protected functions `_add` and `_remove` can be used in `SettingsManagerTest`.

Besides the member variable `Chunk::_indexes`, `Chunk::get_indexes(...)` and `Chunk::get_index(...)` are public functions which can be used to retrieve the created index structures.

**Comment by mweisgut** [comment]
> The operator is supposed to fail because the index with the name already exists

Please check the semantic of `IF NOT EXISTS` in `CREATE INDEX` statements again: The intention of using `IF NOT EXIST` is not to fail when an index with the same name already exists.

>  but the logic error can not be caught for some reason

Have you already debugged the error? To what extent can you narrow down the error? Did you discuss the error with your fellow team members? Could your fellow team members help you with finding the fault?

> we have access to the context and can roll it back, which is okay at that level, but this should happen automatically when working with the high-level SQL pipeline, shouldn't it?

If the operation cannot be executed, Hyrise shall fail loudly. See the Hyrise contribution guidelines (https://github.com/hyrise/hyrise/blob/master/CONTRIBUTING.md#the-four-commandments)

File: src/test/lib/ddl/ddl_statement_test.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  auto chunk_count = targeted_table->chunk_count();
+  for(ChunkID id=ChunkID{0}; id < chunk_count; id+=1) {
+    auto current_chunk = targeted_table->get_chunk(id);
+    auto applied_indices = current_chunk->get_indexes(*column_ids);
+    EXPECT_TRUE(applied_indices.size() == 1);
+  }
+}
+
+TEST_F(DDLStatementTest, CreateIndexIfNotExistsRun) {
+  auto sql_pipeline = SQLPipelineBuilder{"CREATE INDEX myindex IF NOT EXISTS ON table_a (a)"}.create_pipeline();
```
Comment (mweisgut) [inline src/test/lib/ddl/ddl_statement_test.cpp]: This is actually incorrect syntax. The correct syntax would be the following:
`CREATE INDEX IF NOT EXISTS myindex ON table_a (a, b)`

https://www.postgresql.org/docs/current/sql-createindex.html

File: src/test/lib/ddl/ddl_statement_test.cpp (hunk 2 above)
Comment (simonstadlinger) [reply src/test/lib/ddl/ddl_statement_test.cpp]: The check I was referring to in my answer is not the table index statistics check. If you take a look at the last state of this class in the create index branch, we already use the get_index methods of the chunk. But the friends class is also a good hint, thank you.

**Comment by simonstadlinger** [comment]
Hey @mweisgut, I once again reach out to you concerning some tests. Could you please have a look at the implementation in the `table.hpp` file line 224-233. These are three possibilities of how to "fail loudly", if the index we want to drop does not exist and no "IF EXISTS" flag was set in the ddl statement. This case we want to test. The tests, which I do not know how to write, and for which I would ask you how to give some hints on how to test them, are "DropIndexNotExistsNoFlag" in DDLStatementTest and "FailOnWrongIndexName" in DropIndexTest. 

Yes, I have talked to my team members, and I think I am well aware of how this works and should work. I also think its not an error, its just that I don't know how to test this behaviour correctly.

The tests and the code of `table.hpp` should be in the last [commit](https://github.com/simonstadlinger/hyrise/pull/2/commits/25b74b365e42a6e657f3bfd1a24c797d3dbf2b2b) before this comment. 

Thank you for your help!

**Comment by mweisgut** [comment]
> Hey @mweisgut, I once again reach out to you concerning some tests. Could you please have a look at the implementation in the `table.hpp` file line 224-233. These are three possibilities of how to "fail loudly", if the index we want to drop does not exist and no "IF EXISTS" flag was set in the ddl statement. This case we want to test. The tests, which I do not know how to write, and for which I would ask you how to give some hints on how to test them, are "DropIndexNotExistsNoFlag" in DDLStatementTest and "FailOnWrongIndexName" in DropIndexTest.
> 
> Yes, I have talked to my team members, and I think I am well aware of how this works and should work. I also think its not an error, its just that I don't know how to test this behaviour correctly.
> 
> The tests and the code of `table.hpp` should be in the last [commit](https://github.com/simonstadlinger/hyrise/pull/2/commits/25b74b365e42a6e657f3bfd1a24c797d3dbf2b2b) before this comment.
> 
> Thank you for your help!

Hi @simonstadlinger,

Unfortunately, I haven't had the chance yet to take a look at it. If the question is still relevant, I will come back to it in the review phase of the seminar projects.

**Pull request merged**